File: ckl-java/checkerlang/src/ch/checkerlang/values/ValueMap.java
/*  Copyright (c) 2021 Damian Brunold, Gesundheitsdirektion Kanton Zürich

    Permission is hereby granted, free of charge, to any person obtaining a copy
    of this software and associated documentation files (the "Software"), to deal
    in the Software without restriction, including without limitation the rights
    to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
    copies of the Software, and to permit persons to whom the Software is
    furnished to do so, subject to the following conditions:

    The above copyright notice and this permission notice shall be included in all
    copies or substantial portions of the Software.

    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
    SOFTWARE.
*/
package ch.checkerlang.values;

import java.util.Map;
import java.util.TreeMap;

public class ValueMap extends Value {
    private TreeMap<Value, Value> value = new TreeMap<>();

    public ValueMap() {
        // empty
    }

    public ValueMap(Map<Value, Value> value) {
        this.value.putAll(value);
    }

    public ValueMap addItem(Value key, Value value) {
        this.value.put(key, value);
        return this;
    }

    public TreeMap<Value, Value> getValue() {
        return value;
    }

    public boolean isEquals(Value value) {
        if (!value.isMap()) return false;
        TreeMap<Value, Value> other = value.asMap().getValue();
        if (this.value.size() != other.size()) return false;
        for (Value key : this.value.keySet()) {
            if (!other.containsKey(key)) {
                return false;
            }
            if (!this.value.get(key).isEquals(other.get(key))) {
                return false;
            }
        }
        return true;
    }

    public int compareTo(Value value) {
        return toString().compareTo(value.toString());
    }

    public int hashCode() {
        return value.hashCode();
    }

    public String type() {
        return "map";
    }

    public ValueString asString() {
        return new ValueString(toString());
    }

    public ValueInt asInt() {
        return new ValueInt(value.size());
    }

    public ValueBoolean asBoolean() {
        return ValueBoolean.from(value.size() > 0);
    }

    public ValueList asList() {
        ValueList result = new ValueList();
        for (Value item : value.values()) {
            result.addItem(item);
        }
        return result;
    }

    public ValueSet asSet() {
        ValueSet result = new ValueSet();
        for (Value key : value.keySet()) {
            result.addItem(key);
        }
        return result;
    }

    public ValueMap asMap() {
        return this;
    }

    public boolean isMap() {
        return true;
    }

    public String toString() {
        StringBuilder builder = new StringBuilder();
        builder.append("<<<");
        for (Value item : value.keySet()) {
            builder.append(item).append(" => ").append(value.get(item)).append(", ");
        }
        if (builder.length() > "<<<".length()) builder.setLength(builder.length() - 2);
        else if (builder.length() == "<<<".length()) builder.setLength(builder.length() - 1);
        builder.append(">>>");
        return builder.toString();
    }

}

File: ckl-dotnet/ckl-dotnet/ValueMap.cs
/*  Copyright (c) 2021 Damian Brunold, Gesundheitsdirektion Kanton Zürich

    Permission is hereby granted, free of charge, to any person obtaining a copy
    of this software and associated documentation files (the "Software"), to deal
    in the Software without restriction, including without limitation the rights
    to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
    copies of the Software, and to permit persons to whom the Software is
    furnished to do so, subject to the following conditions:

    The above copyright notice and this permission notice shall be included in all
    copies or substantial portions of the Software.

    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
    SOFTWARE.
*/
using System.Collections.Generic;
using System.Text;

namespace CheckerLang
{
    public class ValueMap : Value
    {
        private SortedDictionary<Value, Value> value = new SortedDictionary<Value, Value>(new ComparerValue());

        public ValueMap()
        {
            // empty
        }

        public ValueMap(Dictionary<Value, Value> value)
        {
            foreach (var key in value.Keys)
            {
                this.value[key] = value[key];
            }
        }

        public ValueMap AddItem(Value key, Value value)
        {
            this.value[key] = value;
            return this;
        }

        public SortedDictionary<Value, Value> GetValue()
        {
            return value;
        }


        public override bool IsEquals(Value value)
        {
            if (!value.IsMap()) return false;
            var other = value.AsMap().GetValue();
            if (this.value.Count != other.Count) return false;
            foreach (var key in this.value.Keys)
            {
                if (!other.ContainsKey(key))
                {
                    return false;
                }
                if (!this.value[key].IsEquals(other[key]))
                {
                    return false;
                }
            }
            return true;
        }

        public override int CompareTo(Value value)
        {
            return string.CompareOrdinal(ToString(), value.ToString());
        }

        public override int HashCode()
        {
            return value.GetHashCode();
        }

        public override string Type()
        {
            return "map";
        }

        public override ValueString AsString()
        {
            return new ValueString(ToString());
        }

        public override ValueInt AsInt()
        {
            return new ValueInt(value.Count);
        }

        public override ValueBoolean AsBoolean()
        {
            return ValueBoolean.From(value.Count > 0);
        }

        public override ValueList AsList()
        {
            var result = new ValueList();
            foreach (var item in value.Values)
            {
                result.AddItem(item);
            }
            return result;
        }

        public override ValueSet AsSet()
        {
            var result = new ValueSet();
            foreach (var key in value.Keys)
            {
                result.AddItem(key);
            }
            return result;
        }

        public override ValueMap AsMap()
        {
            return this;
        }

        public override bool IsMap()
        {
            return true;
        }

        public override string ToString()
        {
            var builder = new StringBuilder();
            builder.Append("<<<");
            foreach (var item in value.Keys) {
                builder.Append(item).Append(" => ").Append(value[item]).Append(", ");
            }
            if (builder.Length > "<<<".Length) builder.Remove(builder.Length - ", ".Length, ", ".Length);
            builder.Append(">>>");
            return builder.ToString();
        }

    }
}